File: ex2/main.py
#! /usr/bin/env python3
from threading import Thread, Lock

i = 0
lock = Lock()

def add1(number):
	return number + 1
def sub1(number):
	return number - 1

def thread_function(functor):
	global i
	for j in range(1000000):
		lock.acquire(blocking = True)
		i = functor(i)
		lock.release()

def main():
	thread_1 = Thread(target = thread_function, args = [add1])
	thread_2 = Thread(target = thread_function, args = [sub1])
	thread_1.start()
	thread_2.start()
	thread_1.join()
	thread_2.join()
	print(i)

main()

File: ex2/main.c
#include <pthread.h>
#include <stdio.h>

int i = 0;
pthread_mutex_t mutex;

int add1(int value) {
	return value + 1;
}

int sub1(int value) {
	return value - 1;
}

void thread_function(int (*functor)(int)) {
	for (int j = 0; j < 1000000; ++j) {
		pthread_mutex_lock(&mutex);
		i = functor(i);
		pthread_mutex_unlock(&mutex);
	}
}

int main() {
	pthread_t thread_1, thread_2;
	pthread_create(&thread_1, NULL, (void *(*)(void *)) thread_function, add1);
	pthread_create(&thread_2, NULL, (void *(*)(void *)) thread_function, sub1);
	pthread_join(thread_1, NULL);
	pthread_join(thread_2, NULL);
	printf("%d\n", i);
}
